Merchant page

xmrdsc opened the pull request:
This PR introduces a dedicated merchant page. Current iteration of the merchant features present on the receive page are inadequate to be used as a point of sales system.

The changes in this PR will make it easier for merchants to accept Monero. Run the application in full-screen mode, set the amount to be received and customers scan the QR code.

Alternatively you can provide a payment URL.

Some notes:

- Introduces a menu under Receive called Merchant.
- Optimized for running in full-screen.
- Transaction history shows a 'X time ago' indication. i.e: "the transaction happened 5 minutes ago"
- Incoming transactions are monitored via tx pool.
- For now, clicking on transactions takes you to the History page.
- In the future merchants could optionally specify their to received amount in USD when #1714 is accepted & merged.

![https://i.imgur.com/wjdKQLz.png](https://i.imgur.com/wjdKQLz.png)

![https://i.imgur.com/h1BMDYL.png](https://i.imgur.com/h1BMDYL.png)

[Video](https://i.imgur.com/xk2dlEA.mp4)

xiphon commented:
Would like to keep QR Code and Payment URL field on the Receive tab as it is done in the current master.
Using `Merchant` when you want to receive some funds from a friend looks wrong/misleading to me.

xmrdsc commented:
@xiphon Agreed.

In addition; more feedback from [Reddit](https://www.reddit.com/r/Monero/comments/a4kzf1/feedback_merchant_page_in_the_gui/):

- [ ] If I'm going to display this page to my client, checking "enable sales tracker" should never display my old transactions and amounts! Otherwise my client will know how much the previous client has spent in his dinner.
- [ ] If I'm going to display this page to my client, the disclaimer "it's up to you to accept unconfirmed transactions or not" should appear only the first time this page is opened. My client doesn't need to read this.
- [ ] Amount font should be large.
- [ ] Displaying a Monero logo would be nice.
- [ ] What's preventing the costumer from "tab-ing" back to the regular GUI and exposing the main address, full amount

Will fix these.

dEBRUYNE-1 commented:
Perhaps we could add the QR code to both the receive page and the merchant page?

xmrdsc commented:
@dEBRUYNE-1 

https://i.imgur.com/lMLbkbK.png

xmrdsc commented:
Hiding balance:
![https://i.imgur.com/ZF7euft.gif](https://i.imgur.com/ZF7euft.gif)

Hiding amounts:

![https://i.imgur.com/q7xMEil.gif](https://i.imgur.com/q7xMEil.gif)

xmrdsc commented:
To sum up the changes above:

- Made most text selectable on the merchant page
- Ability to hide the incoming amounts
- Receive page still has QR/Payment URL under 'Advanced Options'
- 'Settings->Layout received a new persistent setting 'Hide Balance'.
- Increased font size for the amount to receive input
- Added Monero logo on the merchant page

xiphon commented:
> Toggle hiding balance via Settings->Layout->Hide balance 

Please submit this in a separate PR

dEBRUYNE-1 commented:
@xmrdsc - I think we should remove the Payment URL and Amount box from the `Receive` page, as I can tell from experience (of helping / assisting users) that it will merely confuse the normal user.

selsta commented:
@dEBRUYNE-1 I see uses for it and it’s under Advanced options anyway, let’s keep the super simple  things for the simple mode :)

xmrdsc commented:
@selsta @dEBRUYNE-1 Middle ground: remove `Amount to receive` - keep QR & Payment URL. Yeah?

xmrdsc commented:
Rebased to latest master.

- Fixed QR code save dialog @selsta 
- Moved commit to hide balance to separate PR #1802 @xiphon 
- Removed 'Amount to receive' from Receive page; [see picture](https://i.imgur.com/eeuod89.png) @dEBRUYNE-1

selsta commented:
I think ideally the receive page would look like this: 
<img width="1081" alt="screenshot 2018-12-13 at 11 15 52" src="https://user-images.githubusercontent.com/7697454/49932104-84d0f080-fec8-11e8-910b-1888e0fbf662.png">
Just Addresses + QR code with no advanced options. The payment URL would be hidden inside the QR code (#1810) and the help text would be updated.

xmrdsc commented:
As the discussion revolves around how the receive page should look like, I'm in favor of reviewing the merchant page and get that in first.

xmrdsc commented:
Rebased.

xiphon commented:
Please squash

xmrdsc commented:
I'd say commit log has already been carefully curated.

erciccione commented:
I agree. With large changes like this one i' d prever more verbosity, therefore, more commits

xiphon commented:
> I agree. With large changes like this one i' d prever more verbosity, therefore, more commits

This is not the case. Each commit must be self-consistent.

xmrdsc commented:
@xiphon I rebased & squashed:

- Add merchant page and components
- Update qml.qrc and .pro

mmbyday commented:
+1 for squashing all
+1 for self-consistent, finished PRs.

xiphon commented:
https://github.com/monero-project/monero-gui/pull/1793#pullrequestreview-186121274 was just an example.

Okay, the next one.

`Prepare MiddlePanel for views with no padding/margins and custom background colors` commit is not self-consistent. It uses `moneroGrey` that doesn't exist at this point.

* Prepare MiddlePanel for views with no padding/margins and custom background colors
* Add Monero grey to style

xmrdsc commented:
Squashed all.

xiphon commented on components/TitleBar.qml:
```diff
@@ -56,6 +56,9 @@ Rectangle {
     property alias titleBarGradientImageOpacity: titleBarGradientImage.opacity
+    property bool orange: false
+    property string buttonHoverColor: "#262626"
+    property string buttonHoverColorOrange: "#44FFFFFF"
```
```suggestion
    property string buttonHoverColorOrange: "#44FFFF"
```

xmrdsc replied:
`#44FFFFFF` is actually correct, try comparing the two.

xmrdsc replied:
http://doc.qt.io/qt-5/qml-color.html

xiphon commented on pages/merchant/Merchant.qml:
```diff
@@ -0,0 +1,697 @@
+                verticalOffset: 3
+                radius: 8.0
+                samples: 16
+                color: "#30000000"
```
Transparent?

xmrdsc replied:
Slightly transparent. Try `#FF000000` and compare.

xiphon replied:
:ok_hand:

xiphon commented on pages/merchant/Merchant.qml:
```diff
@@ -0,0 +1,697 @@
+                verticalOffset: 3
+                radius: 8.0
+                samples: 16
+                color: "#20000000"
```
Same

xmrdsc replied:
Works as intended.

xmrdsc replied:
http://doc.qt.io/qt-5/qml-color.html

xiphon replied:
:ok_hand:

xiphon commented on pages/merchant/Merchant.qml:
```diff
@@ -0,0 +1,697 @@
+                    font.pixelSize: 12 * scaleRatio
+                    font.bold: false
+                    color: "white"
+                    text: "<style type='text/css'>a {text-decoration: none; color: #FF6C3C; font-size: 12px;}</style>Currently selected address: " + addressLabel + " <a href='#'>(Change)</a>"
```
qsTr and place the tags outside

xmrdsc replied:
See previous reply regarding mixing of tags/translatable strings.

xiphon replied:
:ok_hand:

xiphon commented on pages/merchant/Merchant.qml:
```diff
@@ -0,0 +1,697 @@
+                        anchors.fill: parent
+                        hoverEnabled: true
+                        cursorShape: Qt.PointingHandCursor
+                        onClicked:appWindow.showPageRequest("Receive")
```
```suggestion
                        onClicked: appWindow.showPageRequest("Receive")
```

xmrdsc replied:
Fixed.

xiphon commented on pages/TxKey.qml:
```diff
@@ -45,7 +45,7 @@ Rectangle {
     ColumnLayout {
         id: mainLayout
-        anchors.margins: 40 * scaleRatio
+        anchors.margins: (isMobile)? 17 : 20
```
scaleRatio

xmrdsc replied:
Fixed;

```
./pages/Transfer.qml
./pages/Sign.qml
./pages/SharedRingDB.qml
./pages/TxKey.qml
./pages/settings/SettingsWallet.qml
./pages/settings/SettingsLog.qml
./pages/settings/SettingsNode.qml
./pages/settings/SettingsInfo.qml
./pages/settings/SettingsLayout.qml
./pages/Receive.qml
./pages/AddressBook.qml
./pages/Keys.qml
./pages/Mining.qml

```

mmbyday commented on pages/merchant/MerchantCheckbox.qml:
```diff
@@ -0,0 +1,63 @@
+        verticalOffset: 3
+        radius: 8.0
+        samples: 16
+        color: "#20000000"
```
Hex code looks too long.

xmrdsc replied:
As intended, including transparency.

xmrdsc replied:
http://doc.qt.io/qt-5/qml-color.html

mmbyday commented on pages/merchant/MerchantTrackingList.qml:
```diff
@@ -0,0 +1,231 @@
+                        color: "#a8a8a8"
+                        text: {
+                            if(in_txpool){
+                                return qsTr("Awaiting in txpool")
```
```suggestion
                                return qsTr("Awaiting in txpool") + translationManager.emptyString;
```

xmrdsc replied:
Fixed

mmbyday commented on components/TitleBar.qml:
```diff
@@ -70,30 +73,48 @@ Rectangle {
+            anchors.topMargin: 11
+            width: 132
+            height: 22
+            source: "../images/moneroLogo_white.png"
```
have any more of those  2x images? :-)

xmrdsc replied:
I do not :-)

mmbyday commented on components/TitleBar.qml:
```diff
@@ -130,7 +151,13 @@ Rectangle {
             anchors.fill: parent
             cursorShape: Qt.PointingHandCursor
-            onEntered: goToBasicVersionButton.color = "#262626";
+            onEntered: {
```
```suggestion
            onEntered: { goToBasicVersionButton.color = titleBar.orange ? titleBar.buttonHoverColorOrange : titleBar.buttonHoverColor }
```

xmrdsc replied:
Fixed.

xiphon commented on pages/merchant/Merchant.qml:
```diff
@@ -0,0 +1,697 @@
+                        model: trackingModel
+                        message: {
+                            if(!root.enableTracking){
+                                return qsTr(
```
@erciccione what is the best approach from the translator's point of view? I believe strings containing tags may confuse some translators.

xmrdsc replied:
Leaving it as is for now, however I do indeed think it's better to separate the two. Should be part of a different PR, mixed tags/translatable strings are scattered around our codebase.

xiphon commented:
Let's remove `Amount to receive (USD)` for now and add it after #1714 gets merged.

xmrdsc commented:
@xiphon 

> Let's remove Amount to receive (USD) for now and add it after #1714 gets merged.

![](https://i.imgur.com/Msiqklc.png)
